### PR TITLE
Refactor built-in type checking

### PR DESCRIPTION
**Pull request by anderseknert**
Signed-off-by: Anders Eknert <anders@eknert.com>

### CODE REVIEW

File: src/main/clojure/jarl/builtins/array.clj [hunk 1]
```diff
@@ -15,41 +15,27 @@
...
-    (str "unknown type: " (type value) " from value: " value)))
-
-(defn builtin-concat [a b]
-  (when-not (vector? a)
-    (throw (BuiltinException. (format "array.concat: operand 1 must be array but got %s", (java->rego a)))))
-  (when-not (vector? b)
-    (throw (BuiltinException. (format "array.concat: operand 2 must be array but got %s", (java->rego b)))))
+(defn builtin-concat
+  "Implementation of array.concat built-in"
+  {:builtin "array.concat" :args-types ["array" "array"]}
```
Comment (johanfylling) [inline src/main/clojure/jarl/builtins/array.clj:22]: Cool! 😎

File: src/main/clojure/jarl/builtins/array.clj [hunk 2]
```diff
@@ -15,41 +15,27 @@
...
+  [a b]
+  (check-args (meta #'builtin-concat) a b)
```
Comment (johanfylling) [inline src/main/clojure/jarl/builtins/array.clj:24]: Is it possible to pull out the `meta` of func from the outside, where it's being called? If so, we could maybe move this to `jarl.eval/eval-builtin-func`

File: src/main/clojure/jarl/builtins/utils.clj [hunk 3]
```diff
@@ -0,0 +1,33 @@
+(ns jarl.builtins.utils
```
Comment (johanfylling) [inline src/main/clojure/jarl/builtins/utils.clj:1]: Perhaps we should consolidate this with `jarl.util`? Unless we see this as only useful for built-ins.

Comment (johanfylling) [reply src/main/clojure/jarl/builtins/utils.clj:1]: Something to refactor later, perhaps.

**Review by johanfylling — APPROVED** [review]
👍

File: src/main/clojure/jarl/builtins/array.clj (hunk 2 above)
Comment (anderseknert) [reply src/main/clojure/jarl/builtins/array.clj:24]: Yes, that's definitely doable! Sounds reasonable to move the args check there, but I'd need to rewrite the unit tests too for error handling. Will save for a later PR.

File: src/main/clojure/jarl/builtins/utils.clj (hunk 3 above)
Comment (anderseknert) [reply src/main/clojure/jarl/builtins/utils.clj:1]: Yep!

**Pull request merged**